File: clients/kotlin/src/main/kotlin/ai_cup_22/model/Order.kt
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Player's (team's) orders
 */
class Order {
    /**
     * Orders for each of your units
     */
    var unitOrders: MutableMap<Int, ai_cup_22.model.UnitOrder>

    constructor(unitOrders: MutableMap<Int, ai_cup_22.model.UnitOrder>) {
        this.unitOrders = unitOrders
    }

    /**
     * Write Order to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, unitOrders.size)
        for (unitOrdersEntry in unitOrders) {
            val unitOrdersKey = unitOrdersEntry.key
            StreamUtil.writeInt(stream, unitOrdersKey)
            val unitOrdersValue = unitOrdersEntry.value
            unitOrdersValue.writeTo(stream)
        }
    }

    /**
     * Get string representation of Order
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("Order { ")
        stringBuilder.append("unitOrders: ")
        stringBuilder.append(unitOrders)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read Order from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Order {
            var unitOrders: MutableMap<Int, ai_cup_22.model.UnitOrder>
            val unitOrdersSize = StreamUtil.readInt(stream)
            unitOrders = mutableMapOf();
            for (unitOrdersIndex in 0 until unitOrdersSize) {
                var unitOrdersKey: Int
                unitOrdersKey = StreamUtil.readInt(stream)
                var unitOrdersValue: ai_cup_22.model.UnitOrder
                unitOrdersValue = ai_cup_22.model.UnitOrder.readFrom(stream)
                unitOrders.put(unitOrdersKey, unitOrdersValue)
            }
            return Order(unitOrders)
        }
    }
}

File: clients/scala/src/main/scala/ai_cup_22/model/Order.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Player's (team's) orders
 *
 * @param unitOrders Orders for each of your units
 */
case class Order(unitOrders: Map[Int, ai_cup_22.model.UnitOrder]) {
    /**
     * Write Order to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeInt(stream, unitOrders.size)
        unitOrders.foreach { case (key, value) =>
            StreamUtil.writeInt(stream, key)
            value.writeTo(stream)
        }
    }

    /**
     * Get string representation of Order
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Order { ")
        stringBuilder.append("unitOrders: ")
        stringBuilder.append(unitOrders)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Order {
    /**
     * Read Order from input stream
     */
    def readFrom(stream: java.io.InputStream): Order = Order(
        (0 until StreamUtil.readInt(stream)).map { _ => (
            StreamUtil.readInt(stream),
            ai_cup_22.model.UnitOrder.readFrom(stream)
        )}.toMap
    )
}